api/envoy/config/core/v3/address.proto

```diff
@@ -128,6 +128,11 @@ message BindConfig {
   // Additional socket options that may not be present in Envoy source code or
   // precompiled binaries.
   repeated SocketOption socket_options = 3;
+
+  // Additional source addresses append to the `source_address`. This enables
+  // to specify multiple source addresses. The
+  // source address selection is based on the target host address' version.
+  repeated SocketAddress additional_source_addresses = 4;
```

Comment:
AFAICT the only thing this implementation actually does is pick the source address based on v4 vs. v6, right? Like it will pick the first one that matches. Is there ever any case in which it would pick from more than 2 possible addresses? If it's only ever possibly 2, I would make this a single field, and call it 'alternate_source_address` and very clearly document that it's purely based on matching the IP type of the target? I would also do validation and fail of someone supplies 2 v4 or 2 v6 addresses since that could never match.

Let me know if my understanding is wrong, thank you.

Reply:
> AFAICT the only thing this implementation actually does is pick the source address based on v4 vs. v6, right? 

Yes, you are right.

> Like it will pick the first one that matches. Is there ever any case in which it would pick from more than 2 possible addresses? If it's only ever possibly 2, I would make this a single field, and call it 'alternate_source_address` and very clearly document that it's purely based on matching the IP type of the target? I would also do validation and fail of someone supplies 2 v4 or 2 v6 addresses since that could never match.
> 

For dual-stack case, you are right, we only need to two addresses.

From the description of the issue https://github.com/envoyproxy/envoy/issues/9811, there is another use case:
`there are other benefits: as tcp reverse proxy local ip:port is not infinite, consider the TIME_WAIT grace period. By providing more ipv4 addresses in config could mitigate.`

It seems needs more than 2 addresses. If that is a case we may support, then support multiple addresses in the API is more flexiable.

But yes, there is no validation for the dual-stack, people may fill more than 2 addresses, or two same version addresses. If we want to provide validation, we can introduce a field, maybe call 'selectPolicy' when its value set as `version`, then we validate there can be only 2 addresses.

But if we want to support the usecase of 2, I guess in the future the value `selectPolicy` will have `version+fallback`, since people may wants multiple v4 addresses and multiple v6 addresses.

Reply:
OK thanks for the explanation. I guess if we want to future proof the API I would leave it as is in terms of repeated and name, but very clearly clarify what is supported today. Basically that only 1 additional address is supported, it must be of a different IP type then the primary, etc. In the future we could relax this if we want.

Reply:
got it, let me clarify the comment, thanks!